### PR TITLE
Full rework of Ethernet HAL driver

### PR DESCRIPTION
**Pull request by ALABSTM**
The reworked Ethernet HAL driver brings multiple changes and a compatibility break vs. previous Ethernet HAL driver.
Main changes are listed here:

* Add support for PTP and ARP
* Rework packets reception and buffers allocation for better integration and performance
* Rework packets transmission and buffers allocation (using interrupts instead of polling and prevent packets lock during transmission)
* Enhance DMA management
* Decouple Ethernet HAL driver from PHY driver
* Enhance maximum throughput in RX (94Mbs) and TX (92Mbs)
* Enhance footprint (new driver size is 6% less than previous driver)
* Ensure full integration with LwIP (with and without FreeRTOS)
* Rework applications to align with the new Ethernet HAL driver
* Add MISRA-C 2012 compliance, code coverage analysis, static code analysis and robustness validation

### CODE REVIEW

**Review by holden-zenith — APPROVED** [review]

**Comment by Jovan-ux** [comment]
Hi ALABSTM,
Thank you for your Support and provided src. Finally got this integrated. Your rework seems very promising. Thx for examples anyway.
For 2022 for you all the best, KR Jo

File: Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct [hunk 1]
```diff
@@ -11,16 +11,15 @@ LR_IROM1 0x08000000 0x00200000  {    ; load region size_region
...
+  RW_DMARxDscrTab 0x30000000 0x60 {
   *(.RxDecripSection)
   }
-  RW_DMATxDscrTab 0x30040060 0x140 {
+  RW_DMATxDscrTab 0x30000200 0x140 {
   *(.TxDecripSection)
   }
-  RW_Rx_Buffb 0x30040200 0x1800 {
-  *(.RxArraySection)
+  memory_RX_POOL_base 0x30000400 0x12676 {
```
Comment (KnoerleMan) [inline Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct:20]: In the lwipopts.h you set the lwip heap pointer to address 0x3000 4000 (Line 54) and here you use 0x3000 0400.
Is this correct?

Comment (ALABSTM) [reply Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct:20]: Hi @KnoerleMan,

Thank you for your review and for this remark. I will transfer it to our development teams and will get back to you with their answer.

With regards,

**Comment by ALABSTM** [comment]
Hi @holden-zenith,

Thank you for this review.

With regards,

File: Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct (hunk 1 above)
Comment (KnoerleMan) [reply Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct:20]: Hi again,

seems like i maybe understood why the address of the "LwIP_HeapSection" in the scatter file is different from the LWIP_RAM_HEAP_POINTER define in the lwipopts.h file.
The heap section at address 0x30000400 is for the memory pool of the RxBuffers. The required size is about 12K (12 buffers * size of 1000) plus the additional overhead of the RxBuff_t structure.

The heap for the lwip stack (start address = LWIP_RAM_HEAP_POINTER, size = MEM_SIZE) is therefore placed at location 0x30004000 which means the memory pool of the RxBuffers can take up 15360 bytes, which should be enough.
This heap has nothing to do with the scatter file section "LwIP_HeapSection". The lwip heap seems to be used for some Tx operations.

Could someone try to clarify if this is correct?

Comment (ALABSTM) [reply Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct:20]: Hi @KnoerleMan,

You guessed it right! According to our development teams:
* the Rx pool (named `LwIP_HeapSection`) is used for **packets reception** whereas,
* the `LWIP_RAM_HEAP_POINTER` is used for the **LwIP heap**.

The name `LwIP_HeapSection` is clearly confusing and the proposal is to change it into something like `Rx_PoolSection` in both `ethernetif.c` and scatter files. An internal ticket will be logged to have this fixed.

Thank you for having reported this point.

With regards,

Comment (ALABSTM) [reply Projects/NUCLEO-H743ZI/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/STM32H743ZI-Nucleo/STM32H743ZI-Nucleo.sct:20]: ST Internal Reference: 121723

**Comment by MichaelStrohmeier** [comment]
Hi,
if `USE_HAL_ETH_REGISTER_CALLBACKS == 1` is used, the new callbacks (RxAllocateCallback, RxLinkCallback, TxFreeCallback, TxPtpCallback) are not set to their defaults in `ETH_InitCallbacksToDefault`.
 Is this desired?

**Comment by pjukica** [comment]
Hi,
Is the update of all projects which revolve around Ethernet HAL driver planned in this merge request?
There are some examples which aren't updated so I was wondering if that was intentional.
For example, mbedTLS projects which utilize FreeRTOS, LwIP and mbedTLS haven't been updated:
https://github.com/ALABSTM/STM32CubeH7/tree/master/Projects/STM32H743I-EVAL/Applications/mbedTLS

Will this be updated soon?

**Comment by ALABSTM** [comment]
Hi @MichaelStrohmeier and @pjukica,

Thank you for your questions. They have been forwarded to our development teams. We will get back to you as soon as we get their answers.

With regards,

**Comment by ALABSTM** [comment]
Hi @MichaelStrohmeier,

> There are some examples which aren't updated [...] For example, mbedTLS projects which utilize FreeRTOS, LwIP and mbedTLS

Actually, there is no plan to update such applications. These are intended to keep using previous ETH HAL driver which will be moved to both `stm32h7xx_hal_driver\Inc\Legacy` and `stm32h7xx_hal_driver\Src\Legacy` folders.

With regards,

**Comment by ALABSTM** [comment]
Hi @MichaelStrohmeier,

> if USE_HAL_ETH_REGISTER_CALLBACKS == 1 is used, the new callbacks (RxAllocateCallback, RxLinkCallback, TxFreeCallback, TxPtpCallback) are not set to their defaults in ETH_InitCallbacksToDefault. Is this desired?

We confirm this is **not** intended . A fix made available in a future release. Thank you for having reported the point.

With regards,
Ali

**Comment by ALABSTM** [comment]
Hi @ktrofimov,

> But changing send functions from HAL_ETH_Transmit() to HAL_ETH_Transmit_IT() did the thing. Ethernet now works more stable. Thank you for your work!

Glad to read this new ETH HAL driver helped you overcome that issue.

With regards,

**Comment by dima-kapustin** [comment]
I have outgoing UDP traffic about 48 Mb/s and release build fails with HAL 1.10 rather quickly. Debug build runs with no issues, however I cannot say what throughput I get then.

Based on your comment, I may need to try interrupt version of the TX function.

Thanks

**Comment by ktrofimov** [comment]
> overcome that issue

Not really... More tests show that HAL_ETH_Transmit_IT() still fails, but now with the error HAL_ETH_ERROR_BUSY at function call ETH_Prepare_Tx_Descriptors(heth, pTxConfig, 1)

**Comment by ktrofimov** [comment]
Debug shows that no space in PBUF left, but still right after that HAL_ETH_Transmit_IT() got an error HAL_ETH_ERROR_BUSY

**Comment by ktrofimov** [comment]
> I've got an error HAL_ETH_ERROR_BUSY. Only once, but it happened

Google for this problem I did found a possible solution both for Tx and Rx:
```c
static err_t low_level_output(struct netif *netif, struct pbuf *p)
{
    . . .
    res = HAL_ETH_Transmit_IT(&heth, &TxConfig );
    if( res != HAL_OK )
    {
        printf(“%s: HAL_ETH_Transmit failed, errorcode: %lx\n”, __func__, HAL_ETH_GetError(&heth));

        // Clear TBU flag to resume processing
        heth.Instance->DMACSR = ETH_DMACSR_TBU;
        // Instruct the DMA to poll the transmit descriptor list
        heth.Instance->DMACTDTPR = 0;
    }
    . . .
}

and

void ethernetif_input(void* argument)
{
    struct pbuf *p;
    struct netif *netif = (struct netif *) argument;

    for( ;; )
    {
        if (osSemaphoreAcquire(RxPktSemaphore, TIME_WAITING_FOR_INPUT) == osOK)
        {
            do
            {
                p = low_level_input( netif );
                if (p != NULL)
                {
                    if (netif->input( p, netif) != ERR_OK )
                    {
                        pbuf_free(p);
                    }
                }
            } while(p!=NULL);
        }
        if( (heth.Instance->DMACSR & ETH_DMACSR_RBU) != (uint32_t)RESET)
        {
            // Clear RBUS ETHERNET DMA flag
            heth.Instance->DMACSR = ETH_DMACSR_RBU;
            // Resume DMA reception. The register doesn’t care what you write to it.
            heth.Instance->DMACRDTPR = 0;
        }
    }
}
```

**Comment by ALABSTM** [comment]
Hi @dima-kapustin and @ktrofimov,

Thank you for sharing your interesting findings. In case some issue is **confirmed** from your side, please file a new issue or pull-request. This would ease the tracking as this pull-request has been merged and is now considered as closed.

With regards,

**Comment by VanshajKaistha** [comment]
Hi @ALABSTM, 

I am trying to configure my STM32H743XIHx Eval board for ethernet without DHCP and using CMSIS_V2, code was being compiled properly but i am not able to ping my device. I am new to it if anyone can guide me, I will be happy to learn.

Thanks

**Comment by ALABSTM** [comment]
Hi @dima-kapustin and @ktrofimov,

Thank you for sharing your interesting findings. In case some issue is confirmed from your side, **please** do not hesitate to **file a new issue or pull-request**. This would ease the tracking as this pull-request has been merged and is now **closed**.

With regards,

**Comment by ALABSTM** [comment]
Hi @VanshajKaistha,

Would you mind submitting your request to the **ST Community** under the [STM32H7](https://community.st.com/s/topic/0TO0X000000BUxNWAW/stm32h7) topic page. There you might find the support you are looking for.

Indeed, this channel is solely dedicated to topics directly related to the software delivered within this repository, as indicated in the [CONTRIBUTING.md](https://github.com/STMicroelectronics/STM32CubeH7/blob/master/CONTRIBUTING.md#11-before-opening-an-issue) guide. Thank you for your comprehension.

With regards,

**Comment by erwango** [comment]
For interested readers, complementary information could be found here: https://community.st.com/s/question/0D53W00001Gi9BoSAJ/ethernet-hal-driver-reworked-by-st-and-available-in-22q1-preview-now-available-on-github

**Pull request merged**